File: bindings/java-jni/src/main/java/org/maplibre/nativejni/render/FrameScope.java
package org.maplibre.nativejni.render;

/** Tracks callback-scoped access to borrowed native render frame handles. */
final class FrameScope {
  private boolean active = true;

  synchronized void ensureActive() {
    if (!active) {
      throw new IllegalStateException("render frame is no longer active");
    }
  }

  synchronized void close() {
    active = false;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/render/FrameScope.kt
package org.maplibre.nativeffi.render

import kotlin.concurrent.atomics.AtomicInt
import kotlin.concurrent.atomics.ExperimentalAtomicApi

/** Tracks callback-scoped access to borrowed native render frame handles. */
@OptIn(ExperimentalAtomicApi::class)
internal class FrameScope {
  private val active = AtomicInt(1)

  fun ensureActive() {
    check(active.load() != 0) { "render frame is no longer active" }
  }

  fun close() {
    active.store(0)
  }
}
